src/txn/txn_ckpt.c

```diff
@@ -549,6 +552,25 @@ __checkpoint_prepare(WT_SESSION_IMPL *session, bool *trackingp, const char *cfg[
      * which applications can hold open across calls to checkpoint.
      */
     WT_STAT_CONN_SET(session, txn_checkpoint_prep_running, 1);
+
+    /*
+     * It is possible for the stable timestamp to move between the start of our transaction and the
+     * time we acquire the global write lock. To handle this we grab the stable timestamp and use it
+     * later as the checkpoint timestamp. In order to pin the oldest timestamp behind the stable for
+     * the duration that we don't hold the lock we set the read timestamp of the checkpoint,
+     * transaction.
+     */
+    if (use_timestamp) {
+        stable_timestamp = txn_global->stable_timestamp;
+        WT_READ_BARRIER();
```

Comment:
In theory without locking the global txn here we can still have the oldest move ahead of the stable that we acquired here, I can't really see a solution to that that doesn't involve what we were trying to avoid of locking the global txn structure for longer.

Reply:
That's why I thought we might need both solutions under discussion. i.e: a retry and saving a read timestamp to the checkpoint transaction.

On the plus side with this approach - I think the retry probably just needs to be around the choice of a timestamp here and the __wt_txn_begin call, which will probably return `EINVAL`. It will also print a message to the message log, which isn't ideal.
Maybe we can do this without a retry, but it will mean creating a simplified version of:
https://github.com/wiredtiger/wiredtiger/blame/develop/src/txn/txn_timestamp.c#L842-L874

Something like:
```
__wt_readlock(session, &txn_global->rwlock);
txn_shared->read_timestamp = txn_global->stable_timestamp;
WT_ASSERT(session, txn_shared->read_timestamp >= txn_global->pinned_timestamp);
__wt_txn_publish_read_timestamp(session);
__wt_readunlock(session, &txn_global->rwlock);
```

It's probably worth pulling it into a function that lives in src/txn/txn_timestamp.c called something like `__wt_txn_setup_checkpoint_read_timestamp`
I believe that should work if it's done after the call to `__wt_txn_begin` since that matches the API usage of `WT_SESSION::timestamp_transaction("read_timestamp=X")`

Reply:
I don't think the retry is possible because it relies on the stable being moved again. i.e. until that stable timestamp moves you could be looping. It is still possible to have an oldest timestamp set and a stable timestamp of 0, which we could handle.

Additionally if you put the readlock stuff under the call to `txn_begin` doesn't that have the same issue of the stable timestamp drifting between the call to `txn_begin` and you acquiring the read lock. 

The code in `__wt_txn_begin` suggests that we apply the read timestamp and then create the snapshot which would be the correct ordering in my option.

Could we not take a `txn_global` readlock around the call to `__wt_txn_begin` which should handle all the cases. i.e.
```
    if (use_timestamp) {
        /* Take readlock to prevent the oldest timestamp moving forward. */
        __wt_readlock(session, &txn_global->rwlock);
        stable_timestamp = txn_global->stable_timestamp;
        if (stable_timestamp != 0) {
            __wt_timestamp_to_hex_string(stable_timestamp, ts_string);
            WT_RET(__wt_snprintf(
              timestamp_cfg, WT_TS_HEX_STRING_SIZE + 16 + 1, "read_timestamp=%s", ts_string));
            txn_cfg[2] = timestamp_cfg;
        }
    }

    __wt_epoch(session, &conn->ckpt_prep_start);
    WT_RET(__wt_txn_begin(session, txn_cfg));
    if (use_timestamp && stable_timestamp != 0) {
        __wt_readunlock(session, &txn_global->rwlock);
    }
```

Reply:
```I don't think the retry is possible because it relies on the stable being moved again. i.e. until that stable timestamp moves you could be looping. It is still possible to have an oldest timestamp set and a stable timestamp of 0, which we could handle.```

I'm not clear what you are trying to co-ordinate across here. If it isn't making sure that the right stable timestamp is used by the checkpoint, I'm pretty sure we are coding up a solution to a test-only problem here. There is no real relationship between timestamps and the list of active transactions - timestamps visibility doesn't get the benefit of snapshot isolation semantics, and I don't think we'll be able to tweak a check here to make it appear as though they do.